File: test/language/choice_types/java/choice_types/BoolParamChoiceTest.java
package choice_types;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import choice_types.bool_param_choice.BoolParamChoice;

import zserio.runtime.ZserioError;
import zserio.runtime.io.BitBuffer;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

public class BoolParamChoiceTest
{
    @Test
    public void selectorConstructor()
    {
        final boolean selector = true;
        final BoolParamChoice boolParamChoice = new BoolParamChoice(selector);
        assertEquals(selector, boolParamChoice.getSelector());
    }

    @Test
    public void bitStreamReaderConstructor() throws IOException, ZserioError
    {
        final boolean selector = false;
        final byte value = (byte)234;
        final BitBuffer buffer = writeBoolParamChoiceToBitBuffer(selector, value);
        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(buffer.getBuffer(),
                buffer.getBitSize());
        final BoolParamChoice boolParamChoice = new BoolParamChoice(reader, selector);
        assertEquals(selector, boolParamChoice.getSelector());
        assertEquals((short)value, boolParamChoice.getGrey());
    }

    @Test
    public void choiceTag()
    {
        BoolParamChoice boolParamChoice = new BoolParamChoice(true);
        assertEquals(BoolParamChoice.CHOICE_black, boolParamChoice.choiceTag());

        boolParamChoice = new BoolParamChoice(false);
        assertEquals(BoolParamChoice.CHOICE_grey, boolParamChoice.choiceTag());
    }

    @Test
    public void bitSizeOf()
    {
        BoolParamChoice boolParamChoice = new BoolParamChoice(true);
        assertEquals(8, boolParamChoice.bitSizeOf());

        boolParamChoice = new BoolParamChoice(false);
        assertEquals(16, boolParamChoice.bitSizeOf());
    }

    @Test
    public void getSelector()
    {
        final boolean selector = true;
        final BoolParamChoice boolParamChoice = new BoolParamChoice(selector);
        assertEquals(selector, boolParamChoice.getSelector());
    }

    @Test
    public void getSetBlack()
    {
        BoolParamChoice boolParamChoice = new BoolParamChoice(true);
        final byte value = 99;
        boolParamChoice.setBlack(value);
        assertEquals(value, boolParamChoice.getBlack());
    }

    @Test
    public void getSetGrey()
    {
        BoolParamChoice boolParamChoice = new BoolParamChoice(false);
        final short value = 234;
        boolParamChoice.setGrey(value);
        assertEquals(value, boolParamChoice.getGrey());
    }

    @Test
    public void equals()
    {
        BoolParamChoice boolParamChoice1 = new BoolParamChoice(true);
        BoolParamChoice boolParamChoice2 = new BoolParamChoice(true);
        assertTrue(boolParamChoice1.equals(boolParamChoice2));

        final byte value = 99;
        boolParamChoice1.setBlack(value);
        assertFalse(boolParamChoice1.equals(boolParamChoice2));

        boolParamChoice2.setBlack(value);
        assertTrue(boolParamChoice1.equals(boolParamChoice2));

        final byte diffValue = value + 1;
        boolParamChoice2.setBlack(diffValue);
        assertFalse(boolParamChoice1.equals(boolParamChoice2));
    }

    @Test
    public void hashCodeMethod()
    {
        BoolParamChoice boolParamChoice1 = new BoolParamChoice(true);
        BoolParamChoice boolParamChoice2 = new BoolParamChoice(true);
        assertEquals(boolParamChoice1.hashCode(), boolParamChoice2.hashCode());

        final byte value = 99;
        boolParamChoice1.setBlack(value);
        assertTrue(boolParamChoice1.hashCode() != boolParamChoice2.hashCode());

        boolParamChoice2.setBlack(value);
        assertEquals(boolParamChoice1.hashCode(), boolParamChoice2.hashCode());

        final byte diffValue = value + 1;
        boolParamChoice2.setBlack(diffValue);
        assertTrue(boolParamChoice1.hashCode() != boolParamChoice2.hashCode());

        // use hardcoded values to check that the hash code is stable
        assertEquals(31623, boolParamChoice1.hashCode());
        assertEquals(31624, boolParamChoice2.hashCode());
    }

    @Test
    public void initializeOffsets()
    {
        BoolParamChoice boolParamChoice = new BoolParamChoice(true);
        final int bitPosition = 1;
        assertEquals(9, boolParamChoice.initializeOffsets(bitPosition));

        boolParamChoice = new BoolParamChoice(false);
        assertEquals(17, boolParamChoice.initializeOffsets(bitPosition));
    }

    @Test
    public void bitStreamWriterWrite() throws IOException, ZserioError
    {
        boolean selector = true;
        BoolParamChoice boolParamChoice = new BoolParamChoice(selector);
        final byte byteValue = 99;
        boolParamChoice.setBlack(byteValue);
        ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        boolParamChoice.write(writer);
        ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(writer.toByteArray(),
                writer.getBitPosition());
        BoolParamChoice readBoolParamChoice = new BoolParamChoice(reader, selector);
        assertEquals(byteValue, readBoolParamChoice.getBlack());

        selector = false;
        boolParamChoice = new BoolParamChoice(selector);
        final short shortValue = 234;
        boolParamChoice.setGrey(shortValue);
        writer = new ByteArrayBitStreamWriter();
        boolParamChoice.write(writer);
        reader = new ByteArrayBitStreamReader(writer.toByteArray(), writer.getBitPosition());
        readBoolParamChoice = new BoolParamChoice(reader, selector);
        assertEquals(shortValue, readBoolParamChoice.getGrey());
    }

    private BitBuffer writeBoolParamChoiceToBitBuffer(boolean selector, byte value) throws IOException
    {
        try (final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter())
        {
            if (selector)
                writer.writeByte(value);
            else
                writer.writeShort(value);

            return new BitBuffer(writer.toByteArray(), writer.getBitPosition());
        }
    }
}

File: test/language/choice_types/cpp98/BoolParamChoiceTest.cpp
#include "gtest/gtest.h"

#include "choice_types/bool_param_choice/BoolParamChoice.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

namespace choice_types
{
namespace bool_param_choice
{

class BoolParamChoiceTest : public ::testing::Test
{
protected:
    void writeBoolParamChoiceToByteArray(zserio::BitStreamWriter& writer, bool selector, int32_t value)
    {
        if (selector == true)
            writer.writeSignedBits(value, 8);
        else
            writer.writeSignedBits(value, 16);
    }
};

TEST_F(BoolParamChoiceTest, emptyConstructor)
{
    BoolParamChoice boolParamChoice;
    ASSERT_THROW(boolParamChoice.getSelector(), zserio::CppRuntimeException);
}

TEST_F(BoolParamChoiceTest, bitStreamReaderConstructor)
{
    const bool selector = true;
    const int8_t value = 99;
    zserio::BitStreamWriter writer;
    writeBoolParamChoiceToByteArray(writer, selector, value);
    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
    BoolParamChoice boolParamChoice(reader, selector);
    ASSERT_EQ(selector, boolParamChoice.getSelector());
    ASSERT_EQ(value, boolParamChoice.getBlack());
}

TEST_F(BoolParamChoiceTest, copyConstructor)
{
    const bool selector = true;
    BoolParamChoice boolParamChoice;
    boolParamChoice.initialize(selector);
    const int8_t value = 99;
    boolParamChoice.setBlack(value);

    const BoolParamChoice boolParamChoiceCopy(boolParamChoice);
    ASSERT_EQ(selector, boolParamChoiceCopy.getSelector());
    ASSERT_EQ(value, boolParamChoiceCopy.getBlack());
}

TEST_F(BoolParamChoiceTest, operatorAssignment)
{
    const bool selector = false;
    BoolParamChoice boolParamChoice;
    boolParamChoice.initialize(selector);
    const int16_t value = 234;
    boolParamChoice.setGrey(value);

    const BoolParamChoice boolParamChoiceCopy = boolParamChoice;
    ASSERT_EQ(selector, boolParamChoiceCopy.getSelector());
    ASSERT_EQ(value, boolParamChoiceCopy.getGrey());
}

TEST_F(BoolParamChoiceTest, initialize)
{
    const bool selector = false;
    BoolParamChoice boolParamChoice;
    boolParamChoice.initialize(selector);
    ASSERT_EQ(selector, boolParamChoice.getSelector());
}

TEST_F(BoolParamChoiceTest, getSelector)
{
    const bool selector = true;
    BoolParamChoice boolParamChoice;
    boolParamChoice.initialize(selector);
    ASSERT_EQ(selector, boolParamChoice.getSelector());
}

TEST_F(BoolParamChoiceTest, getSetBlack)
{
    BoolParamChoice boolParamChoice;
    boolParamChoice.initialize(true);
    const int8_t value = 99;
    boolParamChoice.setBlack(value);
    ASSERT_EQ(value, boolParamChoice.getBlack());
}

TEST_F(BoolParamChoiceTest, getSetGrey)
{
    BoolParamChoice boolParamChoice;
    boolParamChoice.initialize(false);
    const int16_t value = 234;
    boolParamChoice.setGrey(value);
    ASSERT_EQ(value, boolParamChoice.getGrey());
}

TEST_F(BoolParamChoiceTest, bitSizeOf)
{
    BoolParamChoice boolParamChoiceB;
    boolParamChoiceB.initialize(true);
    ASSERT_EQ(8, boolParamChoiceB.bitSizeOf());

    BoolParamChoice boolParamChoiceG;
    boolParamChoiceG.initialize(false);
    ASSERT_EQ(16, boolParamChoiceG.bitSizeOf());
}

TEST_F(BoolParamChoiceTest, initializeOffsets)
{
    BoolParamChoice boolParamChoiceB;
    boolParamChoiceB.initialize(true);
    const size_t bitPosition = 1;
    ASSERT_EQ(9, boolParamChoiceB.initializeOffsets(bitPosition));

    BoolParamChoice boolParamChoiceG;
    boolParamChoiceG.initialize(false);
    ASSERT_EQ(17, boolParamChoiceG.initializeOffsets(bitPosition));
}

TEST_F(BoolParamChoiceTest, operatorEquality)
{
    BoolParamChoice boolParamChoice1;
    boolParamChoice1.initialize(true);
    BoolParamChoice boolParamChoice2;
    boolParamChoice2.initialize(true);
    ASSERT_TRUE(boolParamChoice1 == boolParamChoice2);

    const int8_t value = 99;
    boolParamChoice1.setBlack(value);
    ASSERT_FALSE(boolParamChoice1 == boolParamChoice2);

    boolParamChoice2.setBlack(value);
    ASSERT_TRUE(boolParamChoice1 == boolParamChoice2);

    const int8_t diffValue = value + 1;
    boolParamChoice2.setBlack(diffValue);
    ASSERT_FALSE(boolParamChoice1 == boolParamChoice2);
}

TEST_F(BoolParamChoiceTest, hashCode)
{
    BoolParamChoice boolParamChoice1;
    boolParamChoice1.initialize(true);
    BoolParamChoice boolParamChoice2;
    boolParamChoice2.initialize(true);
    ASSERT_EQ(boolParamChoice1.hashCode(), boolParamChoice2.hashCode());

    const int8_t value = 99;
    boolParamChoice1.setBlack(value);
    ASSERT_NE(boolParamChoice1.hashCode(), boolParamChoice2.hashCode());

    boolParamChoice2.setBlack(value);
    ASSERT_EQ(boolParamChoice1.hashCode(), boolParamChoice2.hashCode());

    const int8_t diffValue = value + 1;
    boolParamChoice2.setBlack(diffValue);
    ASSERT_NE(boolParamChoice1.hashCode(), boolParamChoice2.hashCode());
}

TEST_F(BoolParamChoiceTest, read)
{
    const bool selector = true;
    BoolParamChoice boolParamChoice;
    boolParamChoice.initialize(selector);

    zserio::BitStreamWriter writer;
    const int8_t value = 99;
    writeBoolParamChoiceToByteArray(writer, selector, value);
    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
    boolParamChoice.read(reader);

    ASSERT_EQ(selector, boolParamChoice.getSelector());
    ASSERT_EQ(value, boolParamChoice.getBlack());
}

TEST_F(BoolParamChoiceTest, write)
{
    bool selector = true;
    BoolParamChoice boolParamChoiceB;
    boolParamChoiceB.initialize(selector);
    const int8_t valueB = 99;
    boolParamChoiceB.setBlack(valueB);
    zserio::BitStreamWriter writerB;
    boolParamChoiceB.write(writerB);
    size_t writeBufferByteSizeB;
    const uint8_t* writeBufferB = writerB.getWriteBuffer(writeBufferByteSizeB);
    zserio::BitStreamReader readerB(writeBufferB, writeBufferByteSizeB);
    BoolParamChoice readBoolParamChoiceB(readerB, selector);
    ASSERT_EQ(valueB, readBoolParamChoiceB.getBlack());

    selector = false;
    BoolParamChoice boolParamChoiceG;
    boolParamChoiceG.initialize(selector);
    const int16_t valueG = 234;
    boolParamChoiceG.setGrey(valueG);
    zserio::BitStreamWriter writerG;
    boolParamChoiceG.write(writerG);
    size_t writeBufferByteSizeG;
    const uint8_t* writeBufferG = writerG.getWriteBuffer(writeBufferByteSizeG);
    zserio::BitStreamReader readerG(writeBufferG, writeBufferByteSizeG);
    BoolParamChoice readBoolParamChoiceG(readerG, selector);
    ASSERT_EQ(valueG, readBoolParamChoiceG.getGrey());
}

} // namespace bool_param_choice
} // namespace choice_types
